### ISSUE TITLE
train_tfdf_spaceship.py

### ISSUE THREAD

**Issue by shawlyn19** [jojogbk19/MOja-MoRaL-#24]
#!/usr/bin/env python3
"""
Train a baseline RandomForest model on the Spaceship Titanic dataset
using TensorFlow Decision Forests.

Usage:
    python train_tfdf_spaceship.py --csv /path/to/train.csv --model_out ./tfdf_rf_spaceship
"""
import argparse
import os
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import tensorflow as tf
import tensorflow_decision_forests as tfdf

def split_cabin(df):
    # Cabin format: "B/0/P" -> Deck / Num / Side
    cabin = df["Cabin"].fillna("Unknown/0/Unknown").astype(str)
    parts = cabin.str.split("/", expand=True)
    # Ensure there are three columns even if split yields fewer
    if parts.shape[1] < 3:
        # pad with Unknown if needed
        for i in range(parts.shape[1], 3):
            parts[i] = "Unknown"
    parts = parts.iloc[:, :3]
    parts.columns = ["CabinDeck", "CabinNum", "CabinSide"]
    # Normalize unknowns
    parts["CabinDeck"] = parts["CabinDeck"].replace("", "Unknown").fillna("Unknown")
    parts["CabinSide"] = parts["CabinSide"].replace("", "Unknown").fillna("Unknown")
    # CabinNum as numeric (coerce errors)
    parts["CabinNum"] = pd.to_numeric(parts["CabinNum"], errors="coerce").fillna(0).astype(int)
    df = df.drop(columns=["Cabin"])
    df = pd.concat([df, parts], axis=1)
    return df

def preprocess(df):
    df = df.copy()

    # Drop identifiers and unused text
    if "PassengerId" in df.columns:
        df = df.drop(columns=["PassengerId"])
    if "Name" in df.columns:
        df = df.drop(columns=["Name"])

    # Split cabin into pieces
    if "Cabin" in df.columns:
        df = split_cabin(df)

    # Normalize boolean-like columns (some CSVs use 'True'/'False' strings)
    for col in ["CryoSleep", "VIP"]:
        if col in df.columns:
            df[col] = df[col].replace(
                {"True": True, "False": False, "true": True, "false": False}
            )
            # Keep NaNs as NaN for now, TFDF can handle missing values.
            # If values are strings other than True/False, convert to "Unknown"
            if df[col].dtype == object:
                df[col] = df[col].fillna("Unknown")

    # Label
    if "Transported" in df.columns:
        # Accept bool or string and convert to int 0/1
        df["Transported"] = df["Transported"].replace(
            {"True": True, "False": False, "true": True, "false": False}
        )
        df["Transported"] = df["Transported"].astype(bool).astype(int)

    # Fill categorical missing values (object dtype)
    cat_cols = df.select_dtypes(include=["object"]).columns.tolist()
    for c in cat_cols:
        df[c] = df[c].fillna("Unknown")

    # Fill numeric missing values with median
    num_cols = df.select_dtypes(include=[np.number]).columns.tolist()
    for n in num_cols:
        df[n] = df[n].fillna(df[n].median())

    return df

def main(args):
    print("Loading dataset:", args.csv)
    df = pd.read_csv(args.csv)
    print("Initial shape:", df.shape)

    df = preprocess(df)
    print("After preprocess shape:", df.shape)

    # Split train/test (stratify by label if available)
    if "Transported" in df.columns:
        X = df.drop(columns=["Transported"])
        y = df["Transported"]
        X_train, X_test, y_train, y_test = train_test_split(
            X, y, test_size=0.2, random_state=42, stratify=y
        )
        train_df = X_train.copy()
        train_df["Transported"] = y_train
        test_df = X_test.copy()
        test_df["Transported"] = y_test
    else:
        # If no label column (e.g., running on test set), just split
        train_df, test_df = train_test_split(df, test_size=0.2, random_state=42)
        y_train = y_test = None

    label = "Transported" if "Transported" in train_df.columns else None
    print("Label column:", label)

    # Create tf datasets
    if label:
        train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(train_df, label=label)
        test_ds = tfdf.keras.pd_dataframe_to_tf_dataset(test_df, label=label)
    else:
        train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(train_df)
        test_ds = None

    # Build the model
    model = tfdf.keras.RandomForestModel(task=tfdf.keras.Task.CLASSIFICATION)
    model.compile(metrics=["accuracy"])  # optional for TFDF

    # Fit
    print("Training RandomForest...")
    model.fit(train_ds)

    # Evaluate
    if test_ds is not None:
        print("Evaluating on test set...")
        eval_res = model.evaluate(test_ds, return_dict=True)
        print("Evaluation:", eval_res)

    # Variable importance / inspector
    inspector = model.make_inspector()
    try:
        vi = inspector.variable_importances()
        if vi:
            print("\nVariable importances (summary):")
            for name, imp in vi.items():
                # Show top 10 for each importance type
                print(f"\n== {name} ==")
                for row in imp[:10]:
                    print(f"  {row.attribute}: {row.importance:.4f}")
        else:
            print("No variable importances available.")
    except Exception as e:
        print("Could not fetch variable importances:", e)

    # Save model
    model_out = args.model_out
    os.makedirs(model_out, exist_ok=True)
    print("Saving model to:", model_out)
    model.save(model_out)
    print("Done.")

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--csv", required=True, help="Path to train.csv")
    parser.add_argument(
        "--model_out", default="./tfdf_rf_spaceship", help="Output folder to save model"
    )
    args = parser.parse_args()
    main(args)
